pulsar-metadata/src/main/java/org/apache/pulsar/metadata/cache/impl/MetadataCacheImpl.java

```diff
@@ -321,22 +325,25 @@ public void accept(Notification t) {
         }
     }
 
-    private CompletableFuture<T> executeWithRetry(Supplier<CompletableFuture<T>> op, String key) {
-        CompletableFuture<T> result = new CompletableFuture<>();
+    private void execute(Supplier<CompletableFuture<T>> op, String key, CompletableFuture<T> result, Backoff backoff) {
         op.get().thenAccept(result::complete).exceptionally((ex) -> {
             if (ex.getCause() instanceof BadVersionException) {
                 // if resource is updated by other than metadata-cache then metadata-cache will get bad-version
                 // exception. so, try to invalidate the cache and try one more time.
                 objCache.synchronous().invalidate(key);
-                op.get().thenAccept(result::complete).exceptionally((ex1) -> {
-                    result.completeExceptionally(ex1.getCause());
-                    return null;
-                });
+                backoffExecutor.schedule(() -> execute(op, key, result, backoff), backoff.next(),
+                        TimeUnit.MILLISECONDS);
                 return null;
             }
             result.completeExceptionally(ex.getCause());
             return null;
         });
+    }
+
+    private CompletableFuture<T> executeWithRetry(Supplier<CompletableFuture<T>> op, String key) {
+        final var backoff = new Backoff(100, TimeUnit.MILLISECONDS, 1, TimeUnit.MINUTES, 0, TimeUnit.MILLISECONDS);
```

Comment:
1 minute might be too long. Generally retrying after a few seconds could never fail, but if there is a bug that makes it retry forever, it could reach the maximum backoff value.

Reply:
In most cases, it should work after two or three tries. If there is an unrecoverable bug, it will cause many conccurent requests in a short time. In this situation, retrying many times quickly is meaningless.

Reply:
After reviewing the details, it seems the retry mechanism is primarily designed to avoid invalidating the cache on the broker. However, I believe this is also incorrectly implemented.

Retries should not occur when another user has updated the data to a new version. Instead, the system should throw a ConcurrentModificationException. This allows users to review the updated data before deciding whether to proceed with their changes. For example, if two users attempt to set the data retention policy for a namespace, one user should receive a ConcurrentModificationException if there is a conflict. Typically, this requires the user to check the latest policy before retrying.

Regarding the incorrect implementation of the retry mechanism, it should verify the data version between the cache and the metadata store. If the versions match, retries should stop, as this indicates the data has been updated elsewhere.

Reply:
In the metadata store API, it seems that we use `BadVersionException` rather than `ConcurrentModificationException` in this case. It means when it tries to update a node in the metadata store, the version is outdated (due to the existing version has already been bumped by another update operation)

Reply:
Now the problem becomes whether we should throw the ConcurrentModificationException/BadVersionException. From a design perspective and according to [the readModifyUpdate documentation](https://github.com/apache/pulsar/blob/216b83008deb469e0fc55ed8117f0c393ebcb0ac/pulsar-metadata/src/main/java/org/apache/pulsar/metadata/api/MetadataCache.java#L108), it seems that MetadataCacheImpl should be responsible for retrying and should not throw this exception.

Since readModifyUpdate does not require users to provide a version, it does not seem reasonable to throw a BadVersionException/ConcurrentModificationException to the caller. If the caller needs to ensure that data is not accidentally overwritten, throwing a BadVersionException/ConcurrentModificationException will not solve this problem. Therefore, unless we can explicitly pass in a version, we cannot achieve this semantic.

Besides, I found that introducing unlimited retries in MetadataCache is also not reasonable. We can use `metadataStoreOperationTimeoutSeconds` as the mandatory stop time for retries and throw the TimeoutException if all retires are failed. Also we can reduce the maximum timeout time following @BewareMyPower 's [suggestion](https://github.com/streamnative/ursa-storage/pull/548#discussion_r1873286488). WDYT? @codelipenghui @BewareMyPower

Reply:
Oh it's right. If we follow the API description, we should add the retry mechanism inside the implementation rather than let users retry manually.

Reply:
> , I found that introducing unlimited retries in MetadataCache is also not reasonable

I think we can provide two options to `MetadataCacheConfig` to specify the backoff and the timeout.

Reply:
> I think we can provide two options to MetadataCacheConfig to specify the backoff and the timeout.

I think we need 3 options: backoffInitialTimeInMs, backoffMaxTimeInMs and backoffMandatoryStopTimeInMs.
And we should use `metadataStoreOperationTimeoutSeconds * 1000` as the default value of `backoffMandatoryStopTimeInMs`.
The default value of `backoffInitialTimeInMs` could be 5ms and the default `backoffMaxTimeInMs` could be 3000ms.

Reply:
Why not just add `Backoff` as its field?

```java
public class MetadataCacheConfig<T> {
    // ...
    private final Backoff backoff = new Backoff(5, TimeUnit.MILLISECONDS, 3, TimeUnit.SECONDS, 0, TimeUnit.MILLISECONDS);
```

Regarding the timeout, I mean if a modification operation is not done after the timeout, just fail the future with `TimeoutException`. It's independent from the `Backoff`

Reply:
> Why not just add Backoff as its field?

OK. That's also fine.

> Regarding the timeout, I mean if a modification operation is not done after the timeout, just fail the future with TimeoutException. It's independent from the Backoff

Should the caller or the metadata cache fail the future? My idea is to let the MetadataCache fail the future with the TimeoutException, so we could use the backoff mandatory stop time as the timeout. But if it's the caller, then it doesn't make sense.  Failing the future this way won't stop the retry task. We could check the future before each retry, but the bigger issue is that the current caller won't fail the future. Without introducing a mandatory stop time, the retry task could run indefinitely.

Reply:
Ah yes, we don't need the timeout. We have `CompletableFuture#orTimeout` to support this purpose. So adding a `Backoff` field is enough

Reply:
I added a new config `retryBackoff` to the MetadataCacheConfig. I also used `isMandatoryStopMade` to check for timeouts. If it times out, the future will be failed.